### ISSUE TITLE
[Bug]: Compression job may leave several chunks uncompressed

### ISSUE THREAD

**Issue by mickael-choisnet** [timescale/timescaledb#5353]
### What type of bug is this?

Incorrect result, Performance issue

### What subsystems and features are affected?

Compression

### What happened?

When inserting data while the compression job is running and recompressing chunks, the job may fault during recompression and continue while leaving the current chunk uncompressed.
Very often we insert new data into compressed chunk and therefore our compression job recompresses a large number of chunks, which results in a few hours of work...
We load new data hourly into the system and each run may contain some old data which will fall into compressed chunks, so inserts always occur during the recompression performed by the job.

**What would I have expected:**
I would expect the job to detect a lock and wait for it to pick it up and do what it has to. Or mark affected chunks to be retried later and not stay uncompressed until next run.

### TimescaleDB version affected

2.9.3

### PostgreSQL version used

PostgreSQL 14.5 (Debian 14.5-1.pgdg100+1)

### What operating system did you use?

Debian GNU/Linux 10 (buster) (10.13)

### What installation method did you use?

Deb/Apt

### What platform did you run on?

Microsoft Azure Cloud

### Relevant log output and stack trace

```bash
2023-02-17 06:43:45 UTC [25188-663] LOG:  process 25188 detected deadlock while waiting for AccessExclusiveLock on relation 3692072478 of database 16402 after 1000.034 ms
2023-02-17 06:43:45 UTC [25188-664] DETAIL:  Processes holding the lock: 25188, 17977, 18031, 19664. Wait queue: .
2023-02-17 06:43:45 UTC [25188-665] CONTEXT:  SQL statement "SELECT public.compress_chunk(chunk_rec.oid)"
	PL/pgSQL function _timescaledb_internal.policy_compression_execute(integer,integer,anyelement,integer,boolean,boolean) line 60 at PERFORM
	SQL statement "CALL _timescaledb_internal.policy_compression_execute(
	        job_id, htid, lag_value::INTERVAL,
	        maxchunks, verbose_log, recompress_enabled
	      )"
	PL/pgSQL function _timescaledb_internal.policy_compression(integer,jsonb) line 51 at CALL
2023-02-17 06:43:46 UTC [25188-666] WARNING:  compressing chunk "_timescaledb_internal._hyper_1_11153_chunk" failed when compression policy is executed
2023-02-17 06:43:46 UTC [25188-667] DETAIL:  Message: (), Detail: ().
2023-02-17 06:43:46 UTC [25188-668] CONTEXT:  PL/pgSQL function _timescaledb_internal.policy_compression_execute(integer,integer,anyelement,integer,boolean,boolean) line 62 at RAISE
	SQL statement "CALL _timescaledb_internal.policy_compression_execute(
	        job_id, htid, lag_value::INTERVAL,
	        maxchunks, verbose_log, recompress_enabled
	      )"
	PL/pgSQL function _timescaledb_internal.policy_compression(integer,jsonb) line 51 at CALL
```

### How can we reproduce the bug?

These steps allow to reproduce the issue as it is a bit heavy and tougher than what we have in our production environment.
Nonetheless it is a plausible scenario as our data are exposed through an API, we load new data hourly and the job takes hours to finish.

1. create table
```sql
    CREATE TABLE public.measures (
        measure_date TIMESTAMP NOT NULL,
        received_date TIMESTAMP NOT NULL,
        value DOUBLE PRECISION,
        id TEXT NOT NULL,
        source SMALLINT NOT NULL,
        device_type SMALLINT NULL,
        device_id TEXT NULL
    );
    SELECT create_hypertable('measures', 'measure_date', chunk_time_interval => INTERVAL '5 DAYS', create_default_indexes => false);
    ALTER TABLE measures SET (
        timescaledb.compress,
        timescaledb.compress_segmentby = 'id'
    );
    SELECT add_compression_policy(hypertable => 'measures', compress_after => INTERVAL '14 DAYS', if_not_exists => true);
    CREATE INDEX ix_measures_id_measure_date_received_date ON public.measures (id, measure_date desc, received_date desc);
```

2. load some data (62 000 031 rows / ~7min)
```sql
    CREATE OR REPLACE PROCEDURE public.load_measures (IN gs_start integer, IN gs_end integer) AS
    $BODY$
    BEGIN
        with recursive days as (
            select '2023-01-01'::timestamp as rdate 
            union all 
            select rdate + INTERVAL '1 day' 
            from days
            where extract(year from rdate + INTERVAL '1 day') = 2023 and extract(month from rdate + INTERVAL '1 day') = 1
        )
        insert into measures(measure_date, received_date, value, id, source, device_type, device_id)
        select
            d.rdate as measure_date, current_timestamp::timestamp as received_date, 
            floor(random() * 10000 + 1000) as value, cast(gs.id as text) as id, 5 as source, 
            42 as device_type, 'ABCDEF' as device_id
        from (select generate_series(gs_start,gs_end) as id) gs
        inner join lateral (select * from days) d on true;
    END;
    $BODY$ 
    LANGUAGE PLPGSQL;

    CALL load_measures(gs_start => 1000000, gs_end => 3000000);
```

3. wait for auto VACUUM to end

4. run compression job (~6min)
```sql
    DO
    $BODY$
        DECLARE
            job_id  INT;		
        BEGIN
            SET client_min_messages TO DEBUG1;

            select job.id into job_id
            from _timescaledb_config.bgw_job job
            inner join _timescaledb_catalog.hypertable hypertable on hypertable.id = job.hypertable_id
            where hypertable.table_name = 'measures';

            CALL run_job(job_id);
        END;
    $BODY$;
```

5. wait for auto VACUUM to end

6. load new data (3 100 031 rows / ~1min)
```sql
    CALL load_measures(gs_start => 3000000, gs_end => 3100000);
```

7. wait for auto VACUUM to end (or not, it will just add more lock/risks)

8. run compression job / load data while job is running / run select queries as well (all in parallel)
```sql
    DO
    $BODY$
        DECLARE
            job_id  INT;		
        BEGIN
            SET client_min_messages TO DEBUG1;

            select job.id into job_id
            from _timescaledb_config.bgw_job job
            inner join _timescaledb_catalog.hypertable hypertable on hypertable.id = job.hypertable_id
            where hypertable.table_name = 'measures';

            CALL run_job(job_id);
        END;
    $BODY$;
```

```sql
    CALL load_measures(gs_start => 3100000, gs_end => 3300000);
```

**Run the queries multiple times if needed to add traffic upon the chunks**

```sql
    select m.*
    from measures m
    inner join json_array_elements_text('["1500000","1550000","1555000","1555500","2500000","2550000","2555000","2555500"]') ids(value) on ids.value = m.id
    where m.measure_date > '2023-01-05' AND m.measure_date < '2023-01-25'
    order by m.id, m.measure_date;
```

```sql
    select m.*
    from measures m
    inner join json_array_elements_text('["1400000","1440000","1444000","1444400","2400000","2440000","2444000","2444400"]') ids(value) on ids.value = m.id
    where m.measure_date > '2023-01-05' AND m.measure_date < '2023-01-25'
    order by m.id, m.measure_date;
```

**Comment by kgyrtkirk** [comment]
* from the above steps I was not able to decide where the incorrect results may have happened; I wonder if there are any incorrect results in this case - as there is another issue in which compression is being interrupted; which leaves some garbage behind; although its not fortunate - it doesn't affect correctness; I wonder if this is a similar issue
* I see that there is no message/detail about the error `policy_compression_execute` encounters;
```
2023-02-17 06:43:46 UTC [25188-667] DETAIL:  Message: (), Detail: ().
```
however there is also:
```
2023-02-17 06:43:45 UTC [25188-663] LOG:  process 25188 detected deadlock while waiting for AccessExclusiveLock on relation 3692072478 of database 16402 after 1000.034 ms
2023-02-17 06:43:45 UTC [25188-664] DETAIL:  Processes holding the lock: 25188, 17977, 18031, 19664. Wait queue: .
2023-02-17 06:43:45 UTC [25188-665] CONTEXT:  SQL statement "SELECT public.compress_chunk(chunk_rec.oid)"
	PL/pgSQL function _timescaledb_internal.policy_compression_execute(integer,integer,anyelement,integer,boolean,boolean) line 60 at PERFORM
	SQL statement "CALL _timescaledb_internal.policy_compression_execute(
	        job_id, htid, lag_value::INTERVAL,
	        maxchunks, verbose_log, recompress_enabled
	      )"
	PL/pgSQL function _timescaledb_internal.policy_compression(integer,jsonb) line 51 at CALL
```
so it seems like there is pretty high demand for that table; but I don't think `policy_compression_execute` could do any harm where this lock timeout happens (neither autovacuum)
* I've tried to reproduce the issue with the above steps - but I'm a bit confused what is the wrong resultset for the last 2 queries

**Comment by choisnetm** [comment]
Hi @kgyrtkirk,

There is no wrong resultset expected from the last 2 queries, they just are select. 
They are provided to create workload upon the table and make it more likely to have the compression job fail and leave one or more chunks uncompressed at the end of the process

The "incorrect result" is that at the end of the process the job is logged as successful but I have uncompressed chunks in the hypertable. Chunks which were compressed before the job runs.
I would like the job to try and run until it performed correctly the task it has to do.

**Comment by xvaara** [comment]
Try on 2.10.0 since the insert for compression is much better.

also if you insert in batches, you could force the background compression job to run after each batch, since 2.10 doesn't compress on insert:
```sql
do $$
declare
  _job_id int;
begin
	select job_id into _job_id from timescaledb_information.jobs where proc_name = 'policy_compression' and hypertable_name = 'data' and hypertable_schema = 'data';
	call run_job(_job_id);
end;
$$;
```
That way the compression is shorter when done after each batch.

**Comment by mickael-choisnet** [comment]
Hi @xvaara,

I missed the release of 2.10.0 but it's nice to hear it's been improved in that regard.
I will try when I find some time to do so.

Your suggestion sounds interesting but I'm not sure why it should be shorter? As long as one chunk has some uncompressed data it will be fully decompressed in order to merge with the new data, then compressed again right? Or is it more complicated than that?

It might be an option to force these jobs to run every night so we don't pile up too much uncompressed data throughout the week.
I made a mistake saying that all of our hourly batches are likely to insert data into compressed chunks. Actually, it is unlikely, but we have a process that runs once a day and this one can and will insert rows (potentially lots) into compressed chunks.

**Comment by mickael-choisnet** [comment]
Sounds wonderfull!! 😍 
I will plan an update to 2.10.0 and some tests as soon as possible.

Thanks.

**Comment by kgyrtkirk** [comment]
Cool! I'll close issue - as the upgrade will help with your issue!
